src/Google.Api.Gax/SourceContext.cs

```diff
@@ -0,0 +1,90 @@
+﻿/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf;
+using ProtoWellKnownTypes = Google.Protobuf.WellKnownTypes;
+
+namespace Google.Api.Gax
+{
+    public class SourceContext
+    {
+        private const string SourceContextFileName = "source-context.json";
+        private ProtoWellKnownTypes.Struct _sourceContext => FindSourceContext();
+        private ProtoWellKnownTypes.Struct _git => _sourceContext?.Fields["git"]?.StructValue;
+        private readonly static Lazy<SourceContext> s_instance = new Lazy<SourceContext>(() => new SourceContext());
+
+        /// <summary>
+        /// Best way to get application folder path.
+        /// http://stackoverflow.com/questions/6041332/best-way-to-get-application-folder-path
+        /// </summary>
+        /// <exception cref="AppDomainUnloadedException">Possibly threw by getting base directory.</exception>
+        private static string SourceContextFilePath => 
+            Path.Combine(AppDomain.CurrentDomain.BaseDirectory, SourceContextFileName);
```

Comment:
Isn't `System.Reflection.Assembly.GetEntryAssembly().Location` a better choice to get the path of the process? It also doesn't need the unofficial `System.AppDomain` package below.